pkg/linter/lintRuleSyntacticRedundantCommon.go

```diff
@@ -0,0 +1,219 @@
+//     If the resulting conn does not contain the connection of the rule r then the block is not redundant
+//     Done? Rule is redundant
+//
+//nolint:gocyclo // better not split into two function
+func findRuleSyntacticRedundant(configs map[string]*vpcmodel.VPCConfig,
+	filterLayerName string) (res []ruleRedundant, err error) {
+	for _, config := range configs {
+		if config.IsMultipleVPCsConfig {
+			continue // no use in executing lint on dummy vpcs
+		}
+		filterLayer := config.GetFilterTrafficResourceOfKind(filterLayerName)
+		rules, err := filterLayer.GetRules()
+		if err != nil {
+			return nil, err
+		}
+		tableToRules, tableToAtomicBlocks := getTableToAtomicBlocks(rules)
+		// iterates over rules, finds those that are redundant (shadowed/implied)
+		for i := range rules {
+			tableIndex := rules[i].Filter.FilterIndex
+			isRedundantRuleIndex := rules[i].RuleIndex
+			tableAtomicBlocks := tableToAtomicBlocks[tableIndex]
+			// gathers atomic blocks within rule's src and dst
+			srcBlocks := getAtomicBlocksOfSrcOrDst(tableAtomicBlocks, rules[i].SrcCidr)
+			dstBlocks := getAtomicBlocksOfSrcOrDst(tableAtomicBlocks, rules[i].DstCidr)
+			// iterates over cartesian product of atomic blocks within rule's src and dst; rule is redundant if all
+			// items in the cartesian product are shadowed/implies
+			ruleIsRedundant := true
+			containRules := map[int]string{}
+			for _, srcAtomicBlock := range srcBlocks {
+				for _, dstAtomicBlock := range dstBlocks {
+					connOfOthers := connection.None()
+					// computes the connection of other/higher priority rules in this atomic point in the 3-dimension space
+					for otherRule := range tableToRules[tableIndex] {
+						// needs to be "other" rule always, for nacl needs to be higher priority rule
+						otherRuleIndex := tableToRules[tableIndex][otherRule].RuleIndex
+						if isRedundantRuleIndex == otherRuleIndex ||
+							(filterLayerName == vpcmodel.NaclLayer &&
+								isRedundantRuleIndex <= otherRuleIndex) {
+							continue
+						}
+						// otherRule contains src, dst and has a relevant connection?
+						if rules[i].IsIngress == tableToRules[tableIndex][otherRule].IsIngress &&
+							srcAtomicBlock.ContainedIn(tableToRules[tableIndex][otherRule].SrcCidr) &&
+							dstAtomicBlock.ContainedIn(tableToRules[tableIndex][otherRule].DstCidr) &&
+							!rules[i].Conn.Intersect(tableToRules[tableIndex][otherRule].Conn).IsEmpty() {
+							connOfOthersOld := connOfOthers
+							connOfOthers = connOfOthers.Union(tableToRules[tableIndex][otherRule].Conn)
+							if !connOfOthersOld.Equal(connOfOthers) {
+								if _, ok := containRules[otherRuleIndex]; !ok {
+									containRules[otherRuleIndex] = tableToRules[tableIndex][otherRule].RuleDesc
+								}
+							}
+						}
+					}
+					// is <src, dst> shadowed/implied by other rules?
+					if !rules[i].Conn.ContainedIn(connOfOthers) {
+						ruleIsRedundant = false
```

Comment:
right. done.

Reply:
Actually not, since we want to present to the user all the rules relevant to the redundancy of the "redundant rule"; this will also require modification in the condition